test/mpi/coll/allred.c

```diff
@@ -412,74 +594,110 @@ static int test_allred(int count, mtest_mem_type_e evenmem, mtest_mem_type_e odd
+        bxor_test_3(mpi_type, category, memtype);
+    }
 
-    test_types_set4(sum, 1);
-    test_types_set4(prod, 1);
+#if MTEST_HAVE_MIN_MPI_VERSION(2,2)
+    int num_complex_types = sizeof(complex_types) / sizeof(MPI_Datatype);
+    for (int i = 0; i < num_byte_types; i++) {
+        MPI_Datatype mpi_type = complex_types[i];
+        int category = CATEGORY_COMPLEX;
+        if (mpi_type != MPI_DATATYPE_NULL) {
+            sum_test_1(mpi_type, category, memtype);
+            prod_test_1(mpi_type, category, memtype);
+        }
+    }
+#endif
 
-    test_types_set5(lor, 1);
-    test_types_set5(lor, 2);
-    test_types_set5(lxor, 1);
-    test_types_set5(lxor, 2);
-    test_types_set5(lxor, 3);
-    test_types_set5(land, 1);
-    test_types_set5(land, 2);
+#if MTEST_HAVE_MIN_MPI_VERSION(2,2) && defined(HAVE__BOOL)
+    int num_logical_types = sizeof(logical_types) / sizeof(MPI_Datatype);
+    for (int i = 0; i < num_byte_types; i++) {
+        MPI_Datatype mpi_type = logical_types[i];
+        int category = CATEGORY_INT;
+        if (mpi_type != MPI_DATATYPE_NULL) {
+            lor_test_1(mpi_type, category, memtype);
+            lor_test_2(mpi_type, category, memtype);
+            lxor_test_1(mpi_type, category, memtype);
+            lxor_test_2(mpi_type, category, memtype);
+            lxor_test_3(mpi_type, category, memtype);
+            land_test_1(mpi_type, category, memtype);
+            land_test_2(mpi_type, category, memtype);
+        }
+    }
+#endif
 
-    maxloc_test(struct int_test, MPI_2INT);
-    maxloc_test(struct long_test, MPI_LONG_INT);
-    maxloc_test(struct short_test, MPI_SHORT_INT);
-    maxloc_test(struct float_test, MPI_FLOAT_INT);
-    maxloc_test(struct double_test, MPI_DOUBLE_INT);
+    int num_pair_types = sizeof(pair_types) / sizeof(MPI_Datatype);
+    for (int i = 0; i < num_pair_types; i++) {
+        maxloc_test(pair_types[i], memtype);
+        minloc_test(pair_types[i], memtype);
+    }
 
-    minloc_test(struct int_test, MPI_2INT);
-    minloc_test(struct long_test, MPI_LONG_INT);
-    minloc_test(struct short_test, MPI_SHORT_INT);
-    minloc_test(struct float_test, MPI_FLOAT_INT);
-    minloc_test(struct double_test, MPI_DOUBLE_INT);
+    free_in_out_sol(memtype);
 
     MPI_Comm_set_errhandler(MPI_COMM_WORLD, MPI_ERRORS_ARE_FATAL);
-    return errs;
 }
 
 int main(int argc, char **argv)
 {
-    int errs = 0;
-
     MTest_Init(&argc, &argv);
 
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+
+    if (size < 2) {
+        fprintf(stderr, "At least 2 processes required\n");
+        MPI_Abort(MPI_COMM_WORLD, 1);
+    }
+
     struct dtp_args dtp_args;
     dtp_args_init(&dtp_args, MTEST_COLL_COUNT, argc, argv);
     while (dtp_args_get_next(&dtp_args)) {
-        errs += test_allred(dtp_args.count, dtp_args.u.coll.evenmem, dtp_args.u.coll.oddmem);
+        count = dtp_args.count;
+        test_allred(dtp_args.u.coll.evenmem, dtp_args.u.coll.oddmem);
```

Comment:
in general i don't really like the conversion of `errs` to a global, but i suppose it is not out of the ordinary in the testsuite so i won't block on it.

Reply:
Yeah, I took the easy way out. Since it is not too much trouble, I'll convert them into returning err count for each test function.